src/module.c

```diff
@@ -7389,6 +7415,184 @@ RedisModuleBlockedClient *moduleBlockClient(RedisModuleCtx *ctx, RedisModuleCmdF
+ * non-zero, all the module's custom auth cbs are unregistered.
+ * Returns 0 otherwise - if the Module has no custom auth cb registered. */
+int moduleSearchOrUnregisterCustomAuthCBs(RedisModule *module, int should_unregister) {
+    int callback_found = 0;
+    listIter li;
+    listNode *ln;
+    listRewind(moduleCustomAuthCallbacks, &li);
+    while ((ln = listNext(&li))) {
+        RedisModuleCustomAuthCtx *ctx = listNodeValue(ln);
+        if (ctx->module == module) {
+            callback_found = 1;
+            if (!should_unregister) return callback_found;
+            listDelNode(moduleCustomAuthCallbacks, ln);
+            zfree(ctx);
+        }
+    }
+    return callback_found;
+}
+
+/* Returns 1 if custom module based auth is in progress on any client. Returns 0 otherwise. */
+int isCustomAuthInProgress() {
+    if (!listLength(moduleCustomAuthCallbacks)) return 0;
+    listIter li;
+    listNode *ln;
+    listRewind(server.clients,&li);
+    while ((ln = listNext(&li)) != NULL) {
+        client *c = listNodeValue(ln);
+        if (clientHasModuleAuthInProgress(c)) {
+            return 1;
+        }
+    }
+    return 0;
+}
+
+/* Search for & attempt next custom auth callback after skipping the ones already attempted.
+ * Returns the result of the custom auth callback. */
+int attemptNextCustomAuthCb(client *c, robj *username, robj *password, const char **err) {
+    int skipped_prev_callbacks = c->custom_auth_ctx == NULL;
+    RedisModuleCustomAuthCtx *cur_auth_ctx = NULL;
+    listNode *ln;
+    listIter li;
+    listRewind(moduleCustomAuthCallbacks, &li);
+    int result = REDISMODULE_AUTH_NOT_HANDLED;
+    while((ln = listNext(&li))) {
+        cur_auth_ctx = listNodeValue(ln);
+        if (skipped_prev_callbacks) {
+            /* Remove the custom auth complete flag before we attempt the next cb. */
+            c->flags &= ~CLIENT_CUSTOM_AUTH_RESULT;
+            RedisModuleCtx ctx;
+            moduleCreateContext(&ctx, cur_auth_ctx->module, REDISMODULE_CTX_NONE);
+            ctx.client = c;
+            *err = NULL;
+            c->custom_auth_ctx = cur_auth_ctx;
+            result = cur_auth_ctx->auth_cb(&ctx, username, password, err);
+            moduleFreeContext(&ctx);
+            /* If the client is blocked, we are still in the auth chain. Set the auth_ctx. */
+            if (c->flags & CLIENT_BLOCKED) {
+                RedisModuleBlockedClient *bc = c->bstate.module_blocked_handle;
+                bc->auth_ctx = cur_auth_ctx;
+            }
+            /* If Auth has not succeeded or failed AND if a blocking auth is not in progress, we try
+             * the next custom auth callback. */
+            if (result == REDISMODULE_AUTH_HANDLED) break;
+        }
+        if (cur_auth_ctx == c->custom_auth_ctx) {
+            skipped_prev_callbacks = 1;
+        }
+    }
+    return result;
+}
+
+/* Helper function to handle a reprocessed unblocked client that had custom auth ongoing.
```

Comment:
```suggestion
/* Helper function to handle a reprocessed unblocked auth client.
```

I still think generally the documentation is confusing because of the reference to "ongoing custom auth". I'm not clear why we aren't saying "blocked on auth" in all those cases.